### PR TITLE
Add DecodeIdx for progressive reconstruction.

### PR DESCRIPTION
**Pull request by klauspost**
Allow doing shard reconstruction - either parity or data - progressively.

Alternative to https://github.com/klauspost/reedsolomon/pull/311

Proposed interface:

```go
	// DecodeIdx allows progressively decoding shards.
	// dst slices that are non-nil will be filled with decoded data.
	// On first call, dst non-nil slices should be zero.
	// expectInput indicates which input shards are expected (true = shard expected).
	// input provides the shards - non-nil slices are treated as provided data.
	// len(dst) must equal len(input) must equal TotalShards.
	// If expectInput == nil (merge mode): input shards are XORed into corresponding dst shards.
	// This allows merging of partial decodings from different sources.
	DecodeIdx(dst [][]byte, expectInput []bool, input [][]byte) error
```

# Progressive decoding

For advanced use cases, you can progressively decode missing shards (data or parity) using `DecodeIdx`. 
This allows you to:

* Reconstruct shards from multiple sources arriving at different times
* Merge partial reconstructions from different nodes in a distributed system
* Incrementally add input shards as they become available

To access progressive decoding, cast your encoder to the `Extensions` interface:

```Go
// Cast to Extensions interface
ext := enc.(reedsolomon.Extensions)
```

## Basic Usage

`DecodeIdx` works with three parameters:
* `dst [][]byte` - Destination slices for reconstructed data (pre-allocated, initially zeroed)
* `expectInput []bool` - Which shards you expect to receive (true = expected)
* `input [][]byte` - The actual input shards for this call

```Go
// Set up reconstruction - we want to reconstruct shards 1 and 4
dst := make([][]byte, totalShards)
dst[1] = make([]byte, shardSize) // Will reconstruct shard 1
dst[4] = make([]byte, shardSize) // Will reconstruct shard 4

// Mark which shards we expect to receive
expectInput := make([]bool, totalShards)
expectInput[0] = true  // We expect shard 0
expectInput[2] = true  // We expect shard 2
expectInput[3] = true  // We expect shard 3
expectInput[5] = true  // We expect shard 5
expectInput[6] = true  // We expect shard 6

// First call - provide some shards
input1 := make([][]byte, totalShards)
input1[0] = shard0data
input1[2] = shard2data

err := ext.DecodeIdx(dst, expectInput, input1)

// Second call - provide remaining shards
input2 := make([][]byte, totalShards)
input2[3] = shard3data
input2[5] = shard5data
input2[6] = shard6data

err = ext.DecodeIdx(dst, expectInput, input2)
// dst[1] and dst[4] now contain the reconstructed data
```

## Progressive Reconstruction Example

```Go
func progressiveReconstruct() {
    // Create encoder: 5 data + 3 parity = 8 total shards
    enc, _ := reedsolomon.New(5, 3)
    ext := enc.(reedsolomon.Extensions)

    // Assume we lost shards 1 and 3, need to reconstruct them
    dst := make([][]byte, 8)
    dst[1] = make([]byte, shardSize)  // Reconstruct shard 1
    dst[3] = make([]byte, shardSize)  // Reconstruct shard 3

    // We need 5 valid shards total - mark which ones we expect
    expectInput := make([]bool, 8)
    expectInput[0] = true  // Have shard 0
    expectInput[2] = true  // Have shard 2
    expectInput[4] = true  // Have shard 4
    expectInput[5] = true  // Have shard 5 (parity)
    expectInput[6] = true  // Have shard 6 (parity)

    // Shards arrive progressively from different sources

    // First source provides shards 0, 2
    input1 := make([][]byte, 8)
    input1[0] = availableShards[0]
    input1[2] = availableShards[2]
    ext.DecodeIdx(dst, expectInput, input1)

    // Second source provides shard 4
    input2 := make([][]byte, 8)
    input2[4] = availableShards[4]
    ext.DecodeIdx(dst, expectInput, input2)

    // Third source provides remaining parity shards
    input3 := make([][]byte, 8)
    input3[5] = availableShards[5]
    input3[6] = availableShards[6]
    ext.DecodeIdx(dst, expectInput, input3)

    // dst[1] and dst[3] now contain the reconstructed data
}
```

## Merging Partial DecodeIdx Results

You can also merge partial reconstructions from different nodes using merge mode (`expectInput == nil`):

```Go
// Node A reconstructs using shards 0,2,4
dstA := make([][]byte, 8)
dstA[1] = make([]byte, shardSize)
// ... perform partial reconstruction on node A

// Node B reconstructs using shards 5,6,7
dstB := make([][]byte, 8)
dstB[1] = make([]byte, shardSize)
// ... perform partial reconstruction on node B

// Merge the results (XOR them together)
err := ext.DecodeIdx(dstA, nil, dstB) // nil = merge mode
// dstA[1] now contains the final reconstructed result
```

## Important Notes

* **Consistency**: `expectInput` must be the same across all calls for a reconstruction
* **Completeness**: You must provide exactly the shards marked in `expectInput`, one time each
* **Automatic shard selection**: When more than `dataShards` positions are marked as `true` in `expectInput`, only the first `dataShards` will be used for reconstruction.
* **Size matching**: All shards must be the same size
* **Zero initialization**: Destination shards should start as zeros on the first call

Progressive decoding is particularly useful for distributed storage systems, network reconstruction scenarios, and cases where input data arrives incrementally.

Performance:

```
BenchmarkDecodeIdx10x2x1M               	   15658	     75896 ns/op	165792.43 MB/s	    2376 B/op	      16 allocs/op
BenchmarkDecodeIdx50x10x1M              	    1125	   1043366 ns/op	60299.62 MB/s	    7211 B/op	      26 allocs/op
BenchmarkDecodeIdx10x2x16M              	     296	   4017567 ns/op	50111.57 MB/s	    2382 B/op	      16 allocs/op
BenchmarkDecodeIdx5x2x1M                	   25806	     47729 ns/op	153785.38 MB/s	    1858 B/op	      15 allocs/op
BenchmarkDecodeIdx10x4x16M              	     310	   3880559 ns/op	60527.63 MB/s	    2510 B/op	      16 allocs/op
BenchmarkDecodeIdx50x20x1M              	    1130	   1062427 ns/op	69087.43 MB/s	    7980 B/op	      27 allocs/op
BenchmarkDecodeIdx10x4x1M               	   15727	     76871 ns/op	190970.63 MB/s	    2504 B/op	      16 allocs/op
BenchmarkDecodeIdxSingleData10x2x1M     	   15615	     76435 ns/op	164621.55 MB/s	    2376 B/op	      16 allocs/op
BenchmarkDecodeIdxSingleParity10x2x1M   	   15672	     76265 ns/op	164988.58 MB/s	    2376 B/op	      16 allocs/op
BenchmarkDecodeIdxSingleData50x10x1M    	    1167	   1028003 ns/op	61200.75 MB/s	    7210 B/op	      26 allocs/op
BenchmarkDecodeIdxSingleParity50x10x1M  	    1142	   1077599 ns/op	58383.99 MB/s	    7210 B/op	      26 allocs/op
BenchmarkDecodeIdxMultiple10x2x1M       	   12396	     92532 ns/op	135984.35 MB/s	    2432 B/op	      18 allocs/op
BenchmarkDecodeIdxMultiple50x10x1M      	     760	   1600039 ns/op	39320.65 MB/s	    9518 B/op	      34 allocs/op
BenchmarkDecodeIdxMultiple10x4x16M      	     219	   5427820 ns/op	43273.55 MB/s	    2634 B/op	      20 allocs/op
BenchmarkDecodeIdxProgressive10x2x1M    	   13236	     87265 ns/op	144191.85 MB/s	    4720 B/op	      32 allocs/op
BenchmarkDecodeIdxProgressive50x10x1M   	    1171	   1058620 ns/op	59430.74 MB/s	   13602 B/op	      48 allocs/op
BenchmarkDecodeIdxProgressive10x4x16M   	     334	   3817446 ns/op	61528.32 MB/s	    4982 B/op	      32 allocs/op
BenchmarkDecodeIdxVsReconstruct10x2x1M/DecodeIdx         	   15786	     76302 ns/op	164909.71 MB/s	    2376 B/op	      16 allocs/op
BenchmarkDecodeIdxVsReconstruct10x2x1M/Reconstruct       	   15806	     76116 ns/op	165312.15 MB/s	    2194 B/op	      11 allocs/op
BenchmarkDecodeIdxVsReconstruct50x10x1M/DecodeIdx        	    1155	   1022275 ns/op	61543.70 MB/s	    7192 B/op	      26 allocs/op
BenchmarkDecodeIdxVsReconstruct50x10x1M/Reconstruct      	    1162	   1023831 ns/op	61450.14 MB/s	    6797 B/op	      17 allocs/op
BenchmarkDecodeIdxParityOnly10x2x1M                      	   13654	     87400 ns/op	143969.90 MB/s	    2432 B/op	      18 allocs/op
BenchmarkDecodeIdxParityOnly50x10x1M                     	     426	   2868809 ns/op	21930.55 MB/s	   12689 B/op	      44 allocs/op

BenchmarkDecodeIdx10x2x1M-16                  	   29215	     39672 ns/op	317171.02 MB/s	    2377 B/op	      16 allocs/op
BenchmarkDecodeIdx50x10x1M-16                 	    2319	    542212 ns/op	116033.14 MB/s	    7201 B/op	      26 allocs/op
BenchmarkDecodeIdx10x2x16M-16                 	     379	   3183222 ns/op	63246.16 MB/s	    2380 B/op	      16 allocs/op
BenchmarkDecodeIdx5x2x1M-16                   	   40719	     28864 ns/op	254293.72 MB/s	    1859 B/op	      15 allocs/op
BenchmarkDecodeIdx10x4x16M-16                 	     376	   3180416 ns/op	73852.29 MB/s	    2523 B/op	      16 allocs/op
BenchmarkDecodeIdx50x20x1M-16                 	    2330	    544591 ns/op	134780.57 MB/s	    7974 B/op	      27 allocs/op
BenchmarkDecodeIdx10x4x1M-16                  	   29872	     41126 ns/op	356950.10 MB/s	    2504 B/op	      16 allocs/op
BenchmarkDecodeIdxSingleData10x2x1M-16        	   29395	     40879 ns/op	307805.84 MB/s	    2376 B/op	      16 allocs/op
BenchmarkDecodeIdxSingleParity10x2x1M-16      	   28665	     42161 ns/op	298448.36 MB/s	    2376 B/op	      16 allocs/op
BenchmarkDecodeIdxSingleData50x10x1M-16       	    2154	    553009 ns/op	113767.69 MB/s	    7203 B/op	      26 allocs/op
BenchmarkDecodeIdxSingleParity50x10x1M-16     	    1933	    558123 ns/op	112725.23 MB/s	    7203 B/op	      26 allocs/op
BenchmarkDecodeIdxMultiple10x2x1M-16          	   25822	     46985 ns/op	267804.21 MB/s	    2432 B/op	      18 allocs/op
BenchmarkDecodeIdxMultiple50x10x1M-16         	    1519	    753398 ns/op	83507.70 MB/s	    9503 B/op	      34 allocs/op
BenchmarkDecodeIdxMultiple10x4x16M-16         	     280	   4316435 ns/op	54415.51 MB/s	    2631 B/op	      20 allocs/op
BenchmarkDecodeIdxProgressive10x2x1M-16       	   20599	     58024 ns/op	216858.74 MB/s	    4720 B/op	      32 allocs/op
BenchmarkDecodeIdxProgressive50x10x1M-16      	    2097	    552520 ns/op	113868.46 MB/s	   13594 B/op	      48 allocs/op
BenchmarkDecodeIdxProgressive10x4x16M-16      	     384	   3062483 ns/op	76696.28 MB/s	    4981 B/op	      32 allocs/op
BenchmarkDecodeIdxVsReconstruct10x2x1M/DecodeIdx-16         	   28321	     41934 ns/op	300065.41 MB/s	    2376 B/op	      16 allocs/op
BenchmarkDecodeIdxVsReconstruct10x2x1M/Reconstruct-16       	   29383	     40703 ns/op	309142.01 MB/s	    2164 B/op	      11 allocs/op
BenchmarkDecodeIdxVsReconstruct50x10x1M/DecodeIdx-16        	    2110	    556478 ns/op	113058.45 MB/s	    7192 B/op	      26 allocs/op
BenchmarkDecodeIdxVsReconstruct50x10x1M/Reconstruct-16      	    2120	    551557 ns/op	114067.14 MB/s	    6386 B/op	      17 allocs/op
BenchmarkDecodeIdxParityOnly10x2x1M-16                      	   26064	     45689 ns/op	275406.42 MB/s	    2432 B/op	      18 allocs/op
BenchmarkDecodeIdxParityOnly50x10x1M-16                     	    1011	   1181333 ns/op	53257.24 MB/s	   12661 B/op	      44 allocs/op
```

## Summary by CodeRabbit

* **New Features**
  * Added a public DecodeIdx API to support progressive, per-index decoding and merging of partial reconstructions.

* **Documentation**
  * Expanded README with Progressive decoding/encoding guides and focused examples demonstrating DecodeIdx/EncodeIdx usage and large-data scenarios.

* **Tests**
  * Added extensive tests and multiple example entry points covering DecodeIdx edge cases, merge/progressive flows, multi-shard reconstruction, verification, and large-data usage.

* **Refactor**
  * Consolidated decoding to reuse a shared decoding-matrix approach.

* **Chores**
  * Backends without per-index decoding now return a clear not-supported error.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: reedsolomon.go [hunk 1]
```diff
@@ -1370,6 +1384,122 @@ func (r *reedSolomon) Reconstruct(shards [][]byte) error {
...
+			return err
+		}
+
+		// Cache the inverted matrix in the tree for future use keyed on the
+		// indices of the invalid rows.
+		err = r.tree.InsertInvertedMatrix(invalidIndices, dataDecodeMatrix, r.totalShards)
+		if err != nil {
+			return err
+		}
+	}
```
[the remark this answers is not in the document]
Comment (klauspost) [reply reedsolomon.go]: The method checks it.

**Comment by cannium** [comment]
So for my [previous example](https://github.com/klauspost/reedsolomon/pull/311#discussion_r2648955729):
```
Node  [1     2     3]     [4     5     6]     [7     8     9]
Shard  D1    D2    P1      D3    D4    P2      D5    D6    P3
```
The process might be:

On node 4(randomly picked from AZ2):

```go
dst := make([][]byte, 9)
dst[0] = make([]byte, shardSize)

az2 := make([][]byte, 9)
az2[3] = D3
az2[4] = D4
az2[5] = P2
ext.DecodeIdx(dst, expectInput, az2)
```
and send dst[0] to node 1.

On node 1:

receive D2, P1, D5, and AZ2(node 4's dst[0]).
```go
dst := make([][]byte, 9)
dst[0] = make([]byte, shardSize)

input := make([][]byte, 9)
input[1] = D2
input[2] = P1
input[6] = D5

ext.DecodeIdx(dst, expectInput, input)

az2 := make([][]byte, 9)
az2[0] = AZ2
ext.DecodeIdx(dst, nil, az2)

```
and now dst[0] is the reconstructed D1? That should work for my use case.

**Comment by Wondertan** [comment]
How hard would it be to implement this feature for leopard16?
